components/MaskedTransition/src/MDCMaskedTransition.m

```diff
@@ -0,0 +1,269 @@
+/*
+ Copyright 2017-present the Material Components for iOS authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCMaskedTransition.h"
+
+#import <MotionAnimator/MotionAnimator.h>
+
+#import "MDCMaskedPresentationController.h"
+#import "MDCMaskedTransitionMotionForContext.h"
+#import "MDCMaskedTransitionMotionSpec.h"
+
+// Math utilities
+
+static inline CGPoint CenterOfFrame(CGRect frame) {
+  return CGPointMake(CGRectGetMidX(frame), CGRectGetMidY(frame));
+}
+
+static inline CGPoint AnchorPointFromPosition(CGPoint position, CGRect bounds) {
+  return CGPointMake(position.x / bounds.size.width, position.y / bounds.size.height);
+}
+
+static inline CGRect FrameCenteredAround(CGPoint position, CGSize size) {
+  return CGRectMake(position.x - size.width / 2,
+                    position.y - size.height / 2,
+                    size.width,
+                    size.height);
+}
+
+static inline CGFloat LengthOfVector(CGVector vector) {
+  return (CGFloat)sqrt(vector.dx * vector.dx + vector.dy * vector.dy);
+}
+
+@interface MDCMaskedTransition () <MDMTransitionWithPresentation, MDMTransitionWithFallback>
+@end
+
+@implementation MDCMaskedTransition {
+  UIView *_sourceView;
+  MDCMaskedPresentationController *_presentationController;
+  BOOL _shouldSlideWhenCollapsed;
+}
+
+- (instancetype)initWithSourceView:(UIView *)sourceView {
+  self = [super init];
+  if (self) {
+    _sourceView = sourceView;
+  }
+  return self;
+}
+
+- (id<MDMTransition>)fallbackTransitionWithContext:(id<MDMTransitionContext>)context {
+  return _shouldSlideWhenCollapsed ? nil : self;
+}
+
+#pragma mark - MDMTransitionWithPresentation
+
+- (UIModalPresentationStyle)defaultModalPresentationStyle {
+  return UIModalPresentationCustom;
+}
+
+- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented
+                                                      presentingViewController:(UIViewController *)presenting
+                                                          sourceViewController:(UIViewController *)source {
+  _presentationController =
+      [[MDCMaskedPresentationController alloc] initWithPresentedViewController:presented
+                                                      presentingViewController:presenting
+                                                 calculateFrameOfPresentedView:_calculateFrameOfPresentedView];
+  return _presentationController;
```

Comment:
It would - the assumption here is that the transition object is only being used with one view controller and can therefor only have one presentation controller.

It may be cleaner if we add a feature to the transition context that exposes the presentation controller if one is available - am happy to add support for that in a follow-up change to Transitioning.